src/directory/managed_directory.rs

```
@@ -207,19 +210,198 @@ impl ManagedDirectory {
         }
         Ok(())
     }
+
+    /// Verify checksum of a managed file
+    pub fn validate_checksum(&self, path: &Path) -> result::Result<bool, OpenReadError> {
+        let reader = self.directory.open_read(path)?;
+        let data = reader.as_slice();
+        let footer = Footer::from_bytes(data);
+        let data = &data[..data.len() - footer.size() as usize];
+        let mut hasher = Hasher::new();
+        hasher.update(data);
+        let crc = hasher.finalize();
+        Ok(crc == footer.crc())
+    }
+
+    /// List files for which checksum does not match content
+    pub fn list_damaged(&self) -> result::Result<HashSet<PathBuf>, OpenReadError> {
+        let mut hashset = HashSet::new();
+        let managed_paths = self
+            .meta_informations
+            .read()
+            .expect("Managed directory rlock poisoned in list damaged.")
+            .managed_paths
+            .clone();
+
+        for path in managed_paths.into_iter() {
+            if !self.validate_checksum(&path)? {
+                hashset.insert(path);
+            }
+        }
+        Ok(hashset)
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum Footer {
+    V0(V0),
+}
+
+impl Footer {
+    pub fn to_bytes(&self) -> Vec<u8> {
+        match self {
+            Footer::V0(f) => f.to_bytes(),
+        }
+    }
+
+    pub fn from_bytes(data: &[u8]) -> Self {
+        let len = data.len();
+        assert!(len >= 2);
+        let size = LittleEndian::read_u16(&data[len - 2..]);
+        assert!(
+            len >= size as usize,
+            "len({}) is smaller than size({})",
+            len,
+            size
+        );
+        let footer = &data[len - size as usize..];
+        let index_version = footer[0];
+        match index_version {
+            0 => Footer::V0(V0::from_bytes(footer)),
+            _ => panic!("unsuported index_version"),
```

Maybe return an error?

---

Can you add a TODO to add a "string version", so that the error message in a backward incompatibility scenario becomes more helpful?